### PR TITLE
Fix drag and drop to allow moving levels to other activity sections

### PR DESCRIPTION
**Pull request by davidsbailey**
## Background

Finishes [PLAT-412] and [PLAT-413]. There were two problems:
1. The bounding boxes are computed for each ActivitySectionCard at the time they are rendered, but they become inaccurate once the user scrolls the page.
2. The logic for moving a LevelToken to a new ActivitySectionCard did not account for the presence of multiple activities.

## Description

This PR solves the problem 1 in the following way:

* on each render, store React refs to each ActivitySectionCard via `setActivitySectionRef`
* on drag start, compute bounding boxes of each ActivitySectionCard via `updateActivitySectionMetrics`
* on drag (on each mousemove while dragging), update the target activity and activity section via `updateTargetActivitySection`

The rest of the logic for actually moving the card remains mostly the same, with some small tweaks to address problem (2) above.

The reason for only computing bounding boxes on drag start, rather than on every mousemove, is for performance. computing bounding boxes can be expensive, because the interact with the DOM. However, one downside of this approach is that if the user scrolls the page while dragging is happening, then the bounding boxes would become invalid again. if we run into this situation, we can deal with it by updating the bounding boxes in response to the scroll event, which should happen much less frequently than the mousemove event.

## Alternatives

There are various other approaches which turned out not to work very well. here are a few notes on each of them:

### alternative: mouse events

instead of using mathematical comparisons and bounding boxes, it seems like we should be able to keep track of when a dragged element moves inside another ActivitySectionCard based on mouse events like mouseEnter, mouseOver or mouseMove. Unfortunately though, while a LevelToken is being dragged, the pointer is usually over the LevelToken, which means all these mouse events bubble up through its current parent, not the one it is being dragged over. Therefore we might never get a mouse event on the desired target ActivitySectionCard.

### alternative: forced update

you can see here that a similar problem for LevelTokens is addressed by forcing React to re-render them, which re-triggers their logic for computing their own bounding boxes: https://github.com/code-dot-org/code-dot-org/blob/bdd84f801ee63f1f51cbee64f9372e67feee8b41/apps/src/lib/levelbuilder/lesson-editor/ActivitySectionCard.jsx#L121-L125

This is a viable approach but is a bit overkill since we do not really need to re-render, we just need to recompute bounding boxes. In the future the forceUpdate approach could be replaced with something more like what's shown in this PR.

## Screenshots

This shows that dragging across activities is working:

drag-and-drop-levels

However, it's worth noting that this minimal example takes up all the vertical space on my 27-inch monitor.  in order for editors to be able to easily move levels around, we will probably need to (1) reduce our usage of horizontal space, (2) make dragging scroll-safe and (3) make dragging scroll the page.

## Testing story

new and existing unit tests on `moveLevelToActivitySection` redux action. No tests covering dragging behavior, though these could be added later as UI tests or possibly apps unit tests.

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

[PLAT-412]: https://codedotorg.atlassian.net/browse/PLAT-412
[PLAT-413]: https://codedotorg.atlassian.net/browse/PLAT-413

### CODE REVIEW

File: apps/test/unit/lib/levelbuilder/lesson-editor/activitiesEditorReduxTest.js [hunk 1]
```diff
@@ -27,7 +27,7 @@ import _ from 'lodash';
 
 const getInitialState = () => ({
   levelKeyList: {},
-  activities: sampleActivities
+  activities: _.cloneDeep(sampleActivities)
```
Comment (dmcavoy) [inline apps/test/unit/lib/levelbuilder/lesson-editor/activitiesEditorReduxTest.js:30]: For my own understanding why do this need to be cloned?

**Review by dmcavoy — APPROVED** [review]
Nice work on this Dave! This looks like it was a tough one to crack but I tested it out locally and it seems to work great. I do think your comments about scrolling being something we might want to address for the future is something worth tracking as I was running into it with just a couple levels and no content entered.

File: apps/test/unit/lib/levelbuilder/lesson-editor/activitiesEditorReduxTest.js (hunk 1 above)
Comment (davidsbailey) [reply apps/test/unit/lib/levelbuilder/lesson-editor/activitiesEditorReduxTest.js:30]: this was a tricky one. without doing the clone here, the sample activities were getting mutated which was causing other tests to break.

**Comment by davidsbailey** [comment]
> I do think your comments about scrolling being something we might want to address for the future is something worth tracking as I was running into it with just a couple levels and no content entered.

Good idea! Tracked here: https://codedotorg.atlassian.net/browse/PLAT-430

**Pull request merged**